Add blog post for CDC (#12025)
buildbuddy-io/buildbuddy: website/blog/content-defined-chunking.md

---
slug: content-defined-chunking
title: "Remote Cache CDC: Reusing Bytes"
description: How content-defined chunking makes remote cache uploads smaller by reusing the bytes that did not change.
authors: tyler-french
date: 2026-04-28:12:00:00
image: /img/blog/cdc_image_socials.png
tags: [product, engineering, performance, bazel]
---

![New content being added to reused content](../static/img/blog/cdc_intro_image.jpg)

_The goal: move the changed bytes, not the whole output._

## Setting the Scene

The next frontier for build caching is not just skipping actions. It is skipping bytes.

Build caching has come a long way. Instead of rebuilding the world after every edit, Bazel and remote caching let teams reuse action outputs across machines and CI jobs. In practice, builds have moved from something closer to O(size of repo) toward O(size of change).

> But "size of change" can be misleading. What really matters is the size of the transitive actions affected by the edit. A single source change can still ripple into many binaries, packages, bundles, and other large outputs.

Reviewer: It may be helpful to be more explicit that even if only a small sequence of bytes changes, that can snowball into thousands of bytes having to be rebuilt. i.e. The issue is not that a lot of code is changing and that's invalidating the build outputs

- They produce large outputs that take up a lot of space in the cache.
- They have a low cache hit rate since they change whenever any transitive input changes.

![Transitive action collapse](../static/img/blog/cdc_transitive_action_collapse.svg)

_A small source change can invalidate the final transitive action._

## So what do we do?

One option would be to make the actions themselves incremental: incremental linking, runtime linking, smarter bundling, smarter packaging, and so on. But this is usually very difficult, and requires extensive changes to the linkers and tools themselves.

And even if we solved that for Go, we would still need separate solutions for C++ linkers, JavaScript bundlers, app packagers, generated archives, and every other action that can produce a large output. That does not scale.

Instead, we can treat this as a generic output problem: these actions create large files, where only a small amount of "content" is changing. With Content-Defined Chunking (CDC), we can leave the actions themselves untouched, while still getting many of the wins of making those actions incremental.

## Content-Defined Chunking

> CDC is a repeatable process to split a file into chunks not based on size, but based on the patterns of content at the boundaries, or more generally, based on their actual contents.

Reviewer: I'd maybe simplify this to just "but based on their contents". The extra precision is probably too technical for the general reader

The TL;DR is: run a rolling hash over a small window of bytes, and split when the hash matches a rare pattern. If you want chunks around 512 KiB on average, choose a pattern that has about a 1 in 512 KiB chance of matching at each byte. If it does not match, shift the window and try again. The boundary is still content-defined, but over time you get the average chunk size you wanted.[^cdc-cut]

> This means that a large file with a few bytes added or removed somewhere in the file usually only changes the nearby chunk(s), and the remaining chunks stay the same.

Reviewer: I'd maybe add something more explicit about how the remaining chunks stay the same because eventually the sliding hash will hit a series of characters who have not changed, so they will be hashed the same way they were the first time. 

Also a simple example might be helpful here to demonstrate how the algorithm works

Author: Updated. Do you think the cdc_chunk_stability.svg is good enough of an example?

<img width="300" height="150" alt="cdc_chunk_stability" src="https://github.com/user-attachments/assets/4478be67-cc01-4e9c-b535-53ad63f12674" />

Reviewer: I think that diagram is good to show how CDC works conceptually, but doesn't show how the algorithm works more technically. Personally I had a hard time understanding how exactly the algorithm "knew" that the surrounding chunks hadn't changed. I think if I saw a detailed example with a rolling hash, it would've been more clear to me

---
slug: content-defined-chunking
title: "Remote Cache CDC: Reusing Bytes"
description: How content-defined chunking makes remote cache uploads smaller by reusing the bytes that did not change.
authors: tyler-french
date: 2026-04-28:12:00:00
image: /img/blog/cdc_image_socials.png
tags: [product, engineering, performance, bazel]
---

![New content being added to reused content](../static/img/blog/cdc_intro_image.jpg)

_The goal: move the changed bytes, not the whole output._

> Remote Cache CDC uses Content-Defined Chunking (CDC) to make large build outputs behave more incrementally. When a binary, bundle, package, or archive is mostly unchanged, BuildBuddy can reuse chunks it has already seen instead of uploading, downloading, or storing the whole file again. In the [Bazel implementation PR](https://github.com/bazelbuild/bazel/pull/28437), this reduced uploaded data and disk cache size by about 40%. To enable it with BuildBuddy, use Bazel 9.1 or 8.7 and pass `--experimental_remote_cache_chunking`.

Reviewer 2: ```suggestion
Remote Cache CDC uses Content-Defined Chunking (CDC) to make large build outputs behave more incrementally. When a binary, bundle, package, or archive is mostly unchanged, BuildBuddy can reuse chunks it has already seen instead of uploading, downloading, or storing the whole file again. In the [Bazel implementation PR](https://github.com/bazelbuild/bazel/pull/28437), this reduced uploaded data and disk cache size by about 40%. To enable it with BuildBuddy, use Bazel 9.1+ (or 8.7+) and pass `--experimental_remote_cache_chunking`.
```

> Remote Cache CDC uses Content-Defined Chunking (CDC) to make large build outputs behave more incrementally. When a binary, bundle, package, or archive is mostly unchanged, BuildBuddy can reuse chunks it has already seen instead of uploading, downloading, or storing the whole file again. In the [Bazel implementation PR](https://github.com/bazelbuild/bazel/pull/28437), this reduced uploaded data and disk cache size by about 40%. To enable it with BuildBuddy, use Bazel 9.1 or 8.7 and pass `--experimental_remote_cache_chunking`.

Reviewer 2: ```suggestion
Remote Cache CDC uses Content-Defined Chunking (CDC) to make large build outputs behave more incrementally. When a binary, bundle, package, or archive is mostly unchanged, BuildBuddy can reuse chunks it has already seen instead of uploading, downloading, or storing the whole file again. In the [Bazel implementation PR](https://github.com/bazelbuild/bazel/pull/28437), this reduced uploaded data and disk cache size by about 40% when benchmarked on BuildBuddy's own repo. To enable it with BuildBuddy, use Bazel 9.1+ or 8.7+ and pass `--experimental_remote_cache_chunking`.
```

_The client only needs chunk offsets, not a second copy of every chunk._

## BuildBuddy Implementation

BuildBuddy supports CDC in the cache service, executors, and cache proxies.

For executor outputs, large blobs can be written as chunks directly. The chunks are stored as normal CAS entries, and the reconstruction info is stored separately.

Our edge proxies and cache proxies can also intercept large reads and writes and translate them into CDC-supported requests. That means clients and executors can keep using normal cache APIs, while the proxy turns large blob transfers into chunked transfers behind the scenes.

This lowers network transfer, lets repeated chunks be skipped, and lets the chunks that do need to move be transferred in parallel. Existing unchunked cache paths still work.

## Results

In production, CDC has already saved hundreds of TiB of cache transfer, and our cache TTL is up because we store less duplicate data.

The [Bazel implementation PR](https://github.com/bazelbuild/bazel/pull/28437) benchmarked 50 commits of the BuildBuddy repo and saw about 40% less data uploaded, about 40% smaller disk cache, and faster builds in that benchmark.

## Closing

> CDC makes remote caching better at what developers actually do all day: make small changes to large codebases that sometimes produce large outputs. Instead of uploading the same bytes again and again, BuildBuddy can reuse the chunks that did not change.

Reviewer 2: ```suggestion
CDC makes remote caching better at what developers actually do all day: make small changes to large codebases that sometimes produce large outputs. Instead of uploading and downloading the same bytes again and again, BuildBuddy (and now Bazel!) can reuse the chunks that did not change, significantly cutting down on cache transfer. Try it today with Bazel 9.1+ or 8.7+, and setting `--experimental_remote_cache_chunking` on your BuildBuddy cache-enabled Bazel builds.
```

---
slug: content-defined-chunking
title: "Remote Cache CDC: Reusing Bytes"
description: How content-defined chunking makes remote cache uploads smaller by reusing the bytes that did not change.
authors: tyler-french
date: 2026-04-30:12:00:00
image: /img/blog/cdc_image_socials.png
tags: [product, engineering, performance, bazel]
---

![New content being added to reused content](../static/img/blog/cdc_intro_image.jpg)

_The goal: move the changed bytes, not the whole output._

> Remote Cache CDC uses Content-Defined Chunking (CDC) to make large build outputs behave more incrementally. When a binary, bundle, package, or archive is mostly unchanged, BuildBuddy can reuse chunks it has already seen instead of re-uploading or re-downloading the entire file.

Reviewer 3: ```suggestion
BuildBuddy's Remote Cache uses Content-Defined Chunking (CDC) to make large build outputs behave more incrementally. When a binary, bundle, package, or archive is mostly unchanged, BuildBuddy can reuse chunks it has already seen instead of re-uploading or re-downloading the entire file.
```

Im not entirely sure about the generic name of "Remote Cache CDC". It kinda saying that this tech is currently available in all remote cache backend, which it does not.

It would be a bit clearer to say: BuildBuddy's Remote Cache uses CDC... to communicate that this is a feature specific to our remote cache right now.

> In the [Bazel implementation PR](https://github.com/bazelbuild/bazel/pull/28437), this reduced uploaded data and disk cache size by about 40% when benchmarked on BuildBuddy's own repo. To enable client-side CDC with BuildBuddy, use Bazel 8.7 or 9.1+ and pass <code className="flag">--experimental_remote_cache_chunking</code>.

Reviewer 3: Nit:

```suggestion
Using [our chunking implementation](https://github.com/bazelbuild/bazel/pull/28437) in Bazel, we observed up-to 40% reduction in data uploaded and disk cache size when benchmarked against our own BuildBuddy repository. To enable client-side CDC with BuildBuddy, use Bazel 8.7 or 9.1+ and pass <code className="flag">--experimental_remote_cache_chunking</code>.
```

a bit more promotional tone, but it does attribute the PR to us.

Author: It could be over 40% in cases. I don't think "up-to" is accurate.

<!-- truncate -->

## Setting the Scene

The next frontier for build caching is not just skipping actions. It is skipping bytes.

Build caching has come a long way. Instead of rebuilding the world after every edit, Bazel and remote caching let teams reuse action outputs across machines and CI jobs. In practice, builds have moved from something closer to O(size of repo) toward O(size of change).

But "size of change" can be misleading. What really matters is the size of the transitive actions affected by the edit. A small source change can still ripple into many binaries, packages, bundles, and other large outputs, even when only a small part of each output actually changes.

> ## Transitive Actions

Reviewer 3: this section reads not very convincing to me because the chain of actions being invalidated together is just the natural of build tools and not the problem we are trying to solve here.

The problem this solves is large blob are not well supported in the remote context: transferring a large blob using bytestream RPC is slow and cache invalidation on large blobs are expensive.

The latter part of this section made it sound like large blob contributed toward the problem, but i think we should make it front and center: large blob support IS the problem we are solving.

Author: Agree that the invalidation chain itself is normal build-system behavior. The problem CDC solves is narrower: when that normal invalidation produces a new large blob that is mostly byte-identical to one the cache already has, whole-blob remote cache transfer and storage are wasteful. I updated the section to make that distinction explicit. Large output size alone is not enough for CDC to help. The output also needs byte reuse. For example, if an output is random or mostly rewritten each time, CDC provides little benefit.

)
```

That second shape is where small source changes can fan out into large output changes. The source edit might only change a small sequence of bytes in the final output, but the output digest is still new.

Without CDC, the cache treats that as a completely new blob, even when most of the binary, bundle, package, or archive is byte-for-byte identical to the previous version. If many final outputs depend on that changed input, they can all get new digests.

For remote caching, this creates two problems:

- Each cache miss is expensive because the outputs are large.
- Misses happen more often because any transitive input change can produce a new output digest.

One workaround is to disable remote caching for these actions. That avoids uploading huge outputs when the expected cache hit is not worth the write cost, but it creates a different problem: the action now has to run every time. It can also make the action harder to move to remote execution, because RBE depends on moving action inputs and outputs efficiently.

So the build avoids one expensive cache write, but gives up reuse entirely.

![Transitive action collapse](../static/img/blog/cdc_transitive_action_collapse.svg)

> _A small source change can invalidate the final transitive action._

Reviewer 3: I think one way to talk about how bad the problem is would be using a simple example of a go_libary(foo) being depended on by libary bar1, bar2,... barN, and each of the bar libary comes with their own tests

one invalidation to foo would cause N test binaries to be recompiled and relinked and reupload/redownload etc... for the final TestRunners actions to download and run them.

the most expensive binaries being invalidated out there are test binaries

Author: Added a case study for Go tests

The [Bazel implementation PR](https://github.com/bazelbuild/bazel/pull/28437) benchmarked 50 commits of the BuildBuddy repo and saw about 40% less data uploaded, about 40% smaller disk cache, and faster builds in that benchmark.

BuildBuddy currently applies chunking to blobs larger than 2 MiB. In one test, only about 4.2% of objects were above that threshold, so most blobs are not chunked.

Within that eligible subset, CDC deduplicated about 85% of written bytes. Across all cache traffic, overall savings are typically in the 20 to 40% range.

As a rule of thumb, CDC works best for outputs that are large and byte-stable across revisions. Linking and packaging tend to be good fits, and most large outputs we see reuse most of their bytes. Bundling is also a good fit when the output is not compressed, obfuscated, or randomized.

Compression is not terrible, but it usually causes more churn. Compressed formats like `tar.gz` archives and Docker image layers are often less chunkable because a small input change can rewrite more of the compressed byte stream. The key property is byte-level similarity, not the file extension.

> ## Introducing Split and Splice APIs

Reviewer 3: I think it would be best if we wrap these sections below as an "Implementation" section and provide an overview.

Perhaps something like

# Implementation

To achieve this, we need to make the change in several places.
- First, we worked with the folks from Huawei and Google to introduce new RPCs inside the standardized bazelbuild/remote-apis. This opens the way for different server and client implementations to use the same set of RPC for CDC.
- Then we implemented this spec on BuildBuddy side:
  + Server layer (note avoid mentioning the proxy directly here because we have yet to announce that anywhere)
  + Executor layer
- Then we implemented this on Bazel side via PR ...

Author: Updated

The read path becomes:

1. Call `SplitBlob` to get the chunk layout for a large blob.
2. Check which chunks are already present in the local cache.
3. Download the missing chunks with `Read` or `BatchReadBlobs`.

The write path is the reverse:

1. After producing a large output, the client or executor runs it through the CDC algorithm to compute chunk boundaries and chunk digests.
2. It calls `FindMissingBlobs` to check which chunks the cache is missing.
3. It uploads only the missing chunks with `Write` or `BatchUpdateBlobs`.
4. It calls `SpliceBlob` to store the reconstruction metadata.

With this model, chunks are stored as normal CAS blobs under their own digests. The reconstruction metadata is keyed by the original large blob digest, so future `SplitBlob` calls can start from the digest they already know and discover the chunk layout.

This also helps distribute storage more evenly. Instead of treating one very large object as an indivisible cache entry, the cache can store and serve smaller chunks across the CAS like any other blob.

![SplitBlob and SpliceBlob flows](../static/img/blog/cdc_splitblob_spliceblob_flows.svg)

> _SplitBlob is the read-side API; SpliceBlob is the write-side API._

Reviewer 3: Might worth giving some historical context here about how Split Splice was driven by the folks from Huawei, with review and contribution from us and Google to finally merge after years of review.

(if we wana go back even further, Lulu did some experiments with our storage backend in late 2023 and see real benefits then)

Author: I’m going to skip adding the full history here. The original Split/Splice work had a different shape and this post is focused on the current remote-cache CDC use case. I think the immutable spec links and remote-apis PR links in "Further Reading" give enough attribution without pulling the main post into implementation history.

### Executors

Executors can upload large action outputs as chunks directly. The executor walks outputs normally, uses the negotiated chunking parameters to compute chunk digests for large files, calls `FindMissingBlobs`, and uploads only the missing chunks. The uploader can read the needed byte ranges from the original file and upload chunks concurrently, instead of keeping a second full copy in memory.

### Cache Proxies

Because cache proxies see normal CAS traffic, they can apply CDC transparently to large ByteStream reads and writes. For writes, the proxy can split the incoming stream into chunks, skip chunks that already exist remotely, and register the original blob with `SpliceBlob`. For reads, the proxy can use `SplitBlob` to fetch chunks from its local cache or the backing cache, in parallel where possible, then warm its local cache for the next request.

> This means CDC can help on multiple hops: Bazel client to BuildBuddy, executor to BuildBuddy, and BuildBuddy proxy to backing cache. Native Split/Splice-aware clients get end-to-end chunked transfers, while existing clients can still benefit when traffic flows through a BuildBuddy proxy. Existing unchunked cache paths continue to work.

Reviewer 3: combine "cache service" and "cache proxies" into "server side" i think. avoid mentioning proxy directly until we officially announce it with documentation explaining details.

Reviewer 4: +1

Author: Sounds good - although we do offer cache proxy binaries like we do executors.

> ## Availability

Reviewer 3: Missing Executor version (relevant for self-hosted executor users)
and relevant config for executor.

Bazel support for CDC was introduced in [bazelbuild/bazel#28437](https://github.com/bazelbuild/bazel/pull/28437), and remote cache CDC is available in Bazel 8.7 and 9.1+.

Bazel clients using BuildBuddy can opt in to local client-side upload/download savings with:

```sh
bazel build //... --experimental_remote_cache_chunking
```

> BuildBuddy servers currently have CDC enabled for executor outputs and large files flowing through cache proxies.

Reviewer 4: s/cache proxies/edge caches

## Further Reading and References

Bazel and BuildBuddy:

- [Bazel implementation PR](https://github.com/bazelbuild/bazel/pull/28437)
- [Recommended Bazel performance flags](/blog/debugging-slow-bazel-builds/#1-am-i-using-the-right-set-of-flags)
- [BuildBuddy remote build execution](/docs/remote-build-execution/)
- [Bazel remote caching](https://bazel.build/remote/caching)

Remote APIs:

- [Remote Execution API: SplitBlob and SpliceBlob](https://github.com/bazelbuild/remote-apis/blob/becdd8f9ff811df88a22d3eadd6341753d51d167/build/bazel/remote/execution/v2/remote_execution.proto#L443-L561)
- [Remote APIs PR updating SplitBlob and SpliceBlob](https://github.com/bazelbuild/remote-apis/pull/353)
- [Remote APIs PR adding cache chunking settings](https://github.com/bazelbuild/remote-apis/pull/357)

Content-defined chunking:

- [FastCDC, USENIX ATC 2016](https://www.usenix.org/conference/atc16/technical-sessions/presentation/xia)
- [FastCDC 2020, IEEE TPDS](https://doi.org/10.1109/TPDS.2020.2984632)
- [Rabin fingerprint](https://en.wikipedia.org/wiki/Rabin_fingerprint)
- [Google cdc-file-transfer](https://github.com/google/cdc-file-transfer)

> [^cdc-cut]: One drawback of simple forward-scanning CDC algorithms is that if no cut is found before the maximum chunk size, the algorithm has to force a cut at the maximum size. MaxCDC improves this by scanning ahead and choosing the offset with the maximum rolling hash value, avoiding some undesirable forced cuts. RepMaxCDC applies MaxCDC repeatedly until chunks are within the desired size range. See [buildbarn/go-cdc](https://github.com/buildbarn/go-cdc) and [RepMaxCDC: repeated application of MaxCDC](https://github.com/buildbarn/go-cdc#repmaxcdc-repeated-application-of-maxcdc).

Reviewer 3: I don't think this section is needed as it's not actionable for the reader.
They don't really know how this would translate to actual benchmark and when/where will this be implemented.

I think it's safe to hide it for now and revisit it in a future blog post once we decided to implement this algo
